### DISCUSSION TITLE
Authenticate with Mattermost not working

### DISCUSSION THREAD

**Discussion by ecker00** [outline/outline#4032] [Self Hosting]
I'm not able to get Outline to authenticate with Mattermost.

- Click the "Continue with Mattermost" button
- Fill in username/password in Mattermost
- Redirected back to Outline and see `Authentication failed – we were unable to sign you in at this time. Please try again.`

I tried images `:latest`, `0.64.4` and `0.63.0` which makes me believe it's not a bug but probably my bad config (or documentation?).

Looking at Outline docker container log, it's throwing an error.

```js
{"error":"WHERE parameter \"providerId\" has invalid \"undefined\" value","stack":"Error: WHERE parameter \"providerId\" has invalid \"undefined\" value\n    at PostgresQueryGenerator.whereItemQuery (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:1655:13)\n    at /opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:1646:25\n    at Array.forEach (<anonymous>)\n    at PostgresQueryGenerator.whereItemsQuery (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:1644:35)\n    at PostgresQueryGenerator.getWhereConditions (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:1987:19)\n    at PostgresQueryGenerator.selectQuery (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:932:28)\n    at PostgresQueryInterface.select (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-interface.js:407:59)\n    at Function.findAll (/opt/outline/node_modules/sequelize/lib/model.js:1134:47)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async Function.findOne (/opt/outline/node_modules/sequelize/lib/model.js:1228:12)\n    at async userCreator (/opt/outline/build/server/commands/userCreator.js:33:16)\n    at async accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:60:20)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:83:22)","level":"error","message":"Error during authentication"}
```

I've been trying to look at the source code ([auth/providers/oidc.ts](https://github.com/outline/outline/blob/main/server/routes/auth/providers/oidc.ts)) to understand why `providerId` would be `undefined`, but can't seem to make sense of it. Either variable `domain` is undefined or `profile.sub`, but what config would define those?

Here are my `docker.env` OIDC_ variables:
```sh
URL=https://<Redacted>

OIDC_CLIENT_ID=<Redacted>
OIDC_CLIENT_SECRET=<Redacted>
OIDC_AUTH_URI=https://<Redacted>/oauth/authorize
OIDC_TOKEN_URI=https://<Redacted>/oauth/access_token
OIDC_USERINFO_URI=https://<Redacted>/api/v4/users/me

OIDC_USERNAME_CLAIM=preferred_username
OIDC_DISPLAY_NAME="Mattermost OpenID"
OIDC_SCOPES=openid profile email
```

**Outline**
This is running on a self-hosted MacOS M1 server.

**Comment by ecker00** [comment]
For other people struggling, I switched to **Keycloak** for auth and had no issues.

**Comment by milgaluppang** [comment]
@ecker00 I also recently tested with mattermost and had the same problem. Have you ever solved it?

**Reply by ecker00** [reply]
No. But my conclusion was that it's probably bad idea to make Auth rely on a specific service. What if you don't use Mattermost in 5 years, then you don't want to keep that around just for auth with Outline. — So using a dedicated independent service like Keycloak is preferable.

**Reply by Limezy** [reply]
I use Dex, very simple and efficient !

**Reply by milgaluppang** [reply]
@Limezy  Thank you for your answer, but our company is using mattermost 7.0 so we are testing it, but it doesn't work ㅜ_ㅜ

**Comment by mmospanenko** [comment]
@tommoor could you please take a look? maybe we're using wrong values for 
OIDC_USERNAME_CLAIM / OIDC_SCOPES for Mattermost, etc. Or we should insert something in DB manually (via migrations). Will be great to solve this. Thank you

**Reply by tommoor** [reply]
It seems as though the Mattermost `OIDC_USERINFO_URI` response is not OIDC compatible, this would need to be fixed on their side. I'd rather not include hacks in the Outline codebase for individual auth providers. Perhaps there is another hidden endpoint somewhere that would fulfill the OIDC spec…

**Reply by rtgibbons** [reply]
Just came across this trying to get Discord auth to work. I'm currently getting the same error above. I haven't done extensive debugging, but it appears as that OIDC wants a sub object for a unique identifier. Outline could accomplish a fix by creating a new env for `OIDC_USERNAME_ID` that would default to `sub` but then it'll allow us to change it for oauth2 endpoints that aren't following OIDC standards. This would follow the same approach that's already in place for the `OIDC_DISPLAY_NAME` at https://github.com/outline/outline/blob/main/plugins/oidc/server/auth/oidc.ts#L112-L113

Discords docs show that the User object has an `id` which lines up with how https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims defines `sub` 

> Subject - Identifier for the End-User at the Issuer.

**Reply by DandrewsDev** [reply]
I ran into this same issue attempting to use Discord auth. I tested out creating a new ENV var for providing a field to use for the provider id in place of sub. 

Follows the same approach as OIDC_DISPLAY_NAME and works great with authenticating with discord. I tested by setting mine to OIDC_PROVIDER_ID_CLAIM=id in my .env

```
diff --git a/app.json b/app.json
index a0629745..a6e1c13e 100644
--- a/app.json
+++ b/app.json
@@ -86,6 +86,11 @@
       "value": "preferred_username",
       "required": false
     },
+    "OIDC_PROVIDER_ID_CLAIM": {
+      "description": "Specify which claims to derive provider id information from. Supports any valid JSON path with the JWT payload",
+      "value": "sub",
+      "required": false
+    },
     "OIDC_DISPLAY_NAME": {
       "description": "Display name for OIDC authentication",
       "value": "OpenID",
diff --git a/plugins/oidc/server/auth/oidc.ts b/plugins/oidc/server/auth/oidc.ts
index 55c8234e..fb354fbf 100644
--- a/plugins/oidc/server/auth/oidc.ts
+++ b/plugins/oidc/server/auth/oidc.ts
@@ -111,13 +111,13 @@ if (
           // Default is 'preferred_username' as per OIDC spec.
           const username = get(profile, env.OIDC_USERNAME_CLAIM);
           const name = profile.name || username || profile.username;
+          const providerId = get(profile, env.OIDC_PROVIDER_ID_CLAIM);
 
           if (!name) {
             throw AuthenticationError(
               `Neither a name or username was returned in the profile parameter, but at least one is required.`
             );
           }
-
           const result = await accountProvisioner({
             ip: ctx.ip,
             team: {
@@ -137,7 +137,7 @@ if (
               providerId: domain,
             },
             authentication: {
-              providerId: profile.sub,
+              providerId,
               accessToken,
               refreshToken,
               expiresIn: params.expires_in,
diff --git a/server/env.ts b/server/env.ts
index 75b36e28..232649bc 100644
--- a/server/env.ts
+++ b/server/env.ts
@@ -506,6 +506,12 @@ export class Environment {
   public OIDC_USERNAME_CLAIM =
     process.env.OIDC_USERNAME_CLAIM ?? "preferred_username";
 
+  /**
+   * The OIDC profile field to use as the provider id. The default value is
+   * "sub".
+   */
+  public OIDC_PROVIDER_ID_CLAIM = process.env.OIDC_PROVIDER_ID_CLAIM ?? "sub";
+
   /**
    * A space separated list of OIDC scopes to request. Defaults to "openid
    * profile email".

```

**Comment by antran22** [comment]
If any body run into this problem, I have a patch that is currently working with outline@0.69.2 on Docker

**docker-compose.yaml:**
```yaml
version: "3.8"
services:
  outline:
    image: outlinewiki/outline:0.69.2
    command: sh -c "sed -i '84 i profile.sub = profile.sub || profile.id' /opt/outline/build/plugins/oidc/server/auth/oidc.js && yarn db:migrate --env production-ssl-disabled && yarn start"
```

What I did was to monkey-patch a line `profile.sub = profile.sub || profile.id` so we fill the `sub` field returned from OAuth provider with the `id` field. Obviously @tommoor may have some better solution to add in the code, but this is what working for me.

**Reply by antran22** [reply]
I can see that this is already fixed here: https://github.com/outline/outline/commit/dca64fe84b3713c379aa7aa2239a89106b80a869,

but for now this is a patch for 0.69.2 before the next release

**Comment by hibobmaster** [comment]
Here is my working config:

mattermost:
![image](https://github.com/outline/outline/assets/32976627/565f0a45-0308-4904-99dd-ea5383147a41)

outline:
![image](https://github.com/outline/outline/assets/32976627/116744ee-141c-43ea-8f38-a2b45e165763)

**Comment by mmospanenko** [comment]
It works now, thank you
